gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java

```diff
@@ -0,0 +1,830 @@
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    EasyMock.expect(storageRpcMock.compose(ImmutableList.of(BLOB_INFO2.toPb(), BLOB_INFO3.toPb()),
+        BLOB_INFO1.toPb(), EMPTY_RPC_OPTIONS)).andReturn(BLOB_INFO1.toPb());
+    EasyMock.replay(storageRpcMock);
+    BlobInfo blob = storage.compose(req);
+    assertEquals(BLOB_INFO1, blob);
+  }
+
+  @Test
+  public void testComposeWithOptions() {
+    Storage.ComposeRequest req =
+        Storage.ComposeRequest
+            .builder()
+            .addSource(BLOB_NAME2, BLOB_NAME3)
+            .target(BLOB_INFO1)
+            .targetOptions(BLOB_TARGET_GENERATION, BLOB_TARGET_METAGENERATION).build();
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.expect(optionsMock.retryParams()).andReturn(RetryParams.noRetries());
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    EasyMock.expect(storageRpcMock.compose(ImmutableList.of(BLOB_INFO2.toPb(), BLOB_INFO3.toPb()),
+        BLOB_INFO1.toPb(), BLOB_TARGET_OPTIONS_COMPOSE)).andReturn(BLOB_INFO1.toPb());
+    EasyMock.replay(storageRpcMock);
+    BlobInfo blob = storage.compose(req);
+    assertEquals(BLOB_INFO1, blob);
+  }
+
+  @Test
+  public void testCopy() {
+    Storage.CopyRequest req =
+        Storage.CopyRequest.builder().source(BUCKET_NAME1, BLOB_NAME2).target(BLOB_INFO1).build();
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.expect(optionsMock.retryParams()).andReturn(RetryParams.noRetries());
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    EasyMock.expect(storageRpcMock.copy(BLOB_INFO2.toPb(), EMPTY_RPC_OPTIONS, BLOB_INFO1.toPb(),
+        EMPTY_RPC_OPTIONS)).andReturn(BLOB_INFO1.toPb());
+    EasyMock.replay(storageRpcMock);
+    BlobInfo blob = storage.copy(req);
+    assertEquals(BLOB_INFO1, blob);
+  }
+
+  @Test
+  public void testCopyWithOptions() {
+    Storage.CopyRequest req =
+        Storage.CopyRequest.builder().source(BUCKET_NAME1, BLOB_NAME2)
+            .sourceOptions(BLOB_SOURCE_GENERATION, BLOB_SOURCE_METAGENERATION).target(BLOB_INFO1)
+            .targetOptions(BLOB_TARGET_GENERATION, BLOB_TARGET_METAGENERATION).build();
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.expect(optionsMock.retryParams()).andReturn(RetryParams.noRetries());
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    EasyMock.expect(
+        storageRpcMock.copy(BLOB_INFO2.toPb(), BLOB_SOURCE_OPTIONS_COPY, BLOB_INFO1.toPb(),
+            BLOB_TARGET_OPTIONS_COMPOSE)).andReturn(BLOB_INFO1.toPb());
+    EasyMock.replay(storageRpcMock);
+    BlobInfo blob = storage.copy(req);
+    assertEquals(BLOB_INFO1, blob);
+  }
+
+  @Test
+  public void testReadAllBytes() {
```

Comment:
@aozarov I'll add `testReadAllBytesWithOptions`. 
`reader()` and `writer()` functions build the options map and call the constructors of `BlobReadChannelImpl` and `BlobWriteChannelImpl`. `BlobReadChannel` and `BlobWriteChannel` have no method to get options so to check that they are passed correctly I must either add it or use reflection (which I don't think is a good idea). Can you think of a better way of doing this?